pkt-meta/src/dst_vni_lookup/mod.rs

```diff
@@ -0,0 +1,352 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Open Network Fabric Authors
+
+use left_right::{Absorb, ReadGuard, ReadHandle, WriteHandle, new_from_empty};
+use std::collections::HashMap;
+use tracing::{debug, error, warn};
+
+use lpm::trie::IpPrefixTrie;
+use net::buffer::PacketBufferMut;
+use net::headers::{TryHeaders, TryIp};
+use net::packet::{DoneReason, Packet};
+use net::vxlan::Vni;
+use pipeline::NetworkFunction;
+
+pub mod setup;
+
+#[derive(thiserror::Error, Debug, Clone)]
+pub enum DstVniLookupError {
+    #[error("Error building dst_vni_lookup table: {0}")]
+    BuildError(String),
+}
+
+#[derive(Debug, Clone)]
+pub struct VniTables {
+    tables_by_vni: HashMap<Vni, VniTable>,
+}
+
+impl VniTables {
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            tables_by_vni: HashMap::new(),
+        }
+    }
+}
+
+impl Default for VniTables {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+enum VniTablesChange {
+    UpdateVniTables(VniTables),
+}
+
+impl Absorb<VniTablesChange> for VniTables {
+    fn absorb_first(&mut self, change: &mut VniTablesChange, _: &Self) {
+        match change {
+            VniTablesChange::UpdateVniTables(vni_tables) => {
+                *self = vni_tables.clone();
+            }
+        }
+    }
+    fn drop_first(self: Box<Self>) {}
+    fn sync_with(&mut self, first: &Self) {
+        *self = first.clone();
+    }
+}
+
+#[derive(Debug)]
+pub struct VniTablesReader(ReadHandle<VniTables>);
+impl VniTablesReader {
+    fn enter(&self) -> Option<ReadGuard<'_, VniTables>> {
+        self.0.enter()
+    }
+}
+
+#[derive(Debug)]
+pub struct VniTablesWriter(WriteHandle<VniTables, VniTablesChange>);
+impl VniTablesWriter {
+    #[must_use]
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> VniTablesWriter {
+        let (w, _r) = new_from_empty::<VniTables, VniTablesChange>(VniTables::new());
+        VniTablesWriter(w)
+    }
+    #[must_use]
+    pub fn get_reader(&self) -> VniTablesReader {
+        VniTablesReader(self.0.clone())
+    }
+    pub fn update_vni_tables(&mut self, vni_tables: VniTables) {
+        self.0.append(VniTablesChange::UpdateVniTables(vni_tables));
+        self.0.publish();
+        debug!("Updated tables for Destination VNI Lookup");
+    }
+}
+
+#[derive(Debug, Clone)]
+struct VniTable {
+    dst_vnis: IpPrefixTrie<Vni>,
+}
+
+impl VniTable {
+    fn new() -> Self {
+        Self {
+            dst_vnis: IpPrefixTrie::new(),
+        }
+    }
+}
+
+impl Default for VniTable {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub struct DstVniLookup {
+    name: String,
+    tablesr: VniTablesReader,
+}
+
+impl DstVniLookup {
```

Comment:
Again, I am not sure this is the best naming for this NF for serveral reasons:
* True, it does a lookup (or more), but the goal is not the lookup but the marking of the packet so that it gets routed to the right VPC. So, the function is more than a lookup but a routing action.
* Also, I would not tie the name to a VNI either, because we may use other "Discriminants" to determine how to route a packet. In that regard, I'd actually not have called the VniTable as such but instead:
     * let the NF be called VPCrouting or the like (no vnis but VPCs here, generic).
     * let the VniTable be a `VpcDiscriminant` table instead.

I'm not willing to block this PR for this, though. I have as a pending task unifying the types used across stages and can address that then.

Reply:
I struggled to name this stage.  I am glad to rename this stage later after we make the discriminant change.

Reply:
(And yes, we should move to discriminant later)